hibernate-core/src/main/java/org/hibernate/sql/results/spi/ListResultsConsumer.java

```diff
@@ -51,35 +62,57 @@ public ListResultsConsumer(boolean uniqueFilter) {
 			final PersistenceContext persistenceContext = session.getPersistenceContext();
 			persistenceContext.getLoadContexts().register( jdbcValuesSourceProcessingState );
 
-			boolean uniqueRows = false;
-			final Class<R> resultJavaType = rowReader.getResultJavaType();
-			if ( uniqueFilter && resultJavaType != null && ! resultJavaType.isArray() ) {
-				final EntityPersister entityDescriptor = session.getFactory().getMetamodel().findEntityDescriptor( resultJavaType );
-				if ( entityDescriptor != null ) {
-					uniqueRows = true;
-				}
-			}
-
 			final List<R> results = new ArrayList<>();
 
-			while ( rowProcessingState.next() ) {
-				final R row = rowReader.readRow( rowProcessingState, processingOptions );
-
-				boolean add = true;
-				if ( uniqueRows ) {
-					if ( results.contains( row ) ) {
-						add = false;
+			if ( uniqueSemantic == UniqueSemantic.NONE ) {
+				while ( rowProcessingState.next() ) {
+					results.add( rowReader.readRow( rowProcessingState, processingOptions ) );
+					rowProcessingState.finishRowProcessing();
+				}
+			}
+			else {
+				boolean uniqueRows = false;
+				final Class<R> resultJavaType = rowReader.getResultJavaType();
+				if ( resultJavaType != null && !resultJavaType.isArray() ) {
+					final EntityPersister entityDescriptor = session.getFactory().getMetamodel().findEntityDescriptor(
+							resultJavaType );
+					if ( entityDescriptor != null ) {
+						uniqueRows = true;
```

Comment:
Maybe we could do something like this?

```suggestion
			boolean uniqueRows = false;
			if ( uniqueSemantic != UniqueSemantic.NONE ) {
				final Class<R> resultJavaType = rowReader.getResultJavaType();
				if ( resultJavaType != null && !resultJavaType.isArray() ) {
					final EntityPersister entityDescriptor = session.getFactory().getMetamodel().findEntityDescriptor(
							resultJavaType );
					if ( entityDescriptor != null ) {
						uniqueRows = true;
					}
				}
			}
			if ( uniqueRows ) {
				final List<JavaTypeDescriptor> resultJavaTypeDescriptors = rowReader.getResultJavaTypeDescriptors();
				assert resultJavaTypeDescriptors.size() == 1;
				final JavaTypeDescriptor<R> resultJavaTypeDescriptor = resultJavaTypeDescriptors.get( 0 );
				while ( rowProcessingState.next() ) {
					final R row = rowReader.readRow( rowProcessingState, processingOptions );
					boolean add = true;
				        for ( R existingRow : results ) {
					        if ( resultJavaTypeDescriptor.areEqual( existingRow, row ) ) {
						        if ( uniqueSemantic == UniqueSemantic.ASSERT && !rowProcessingState.hasCollectionInitializers() ) {
							        throw new HibernateException(
									        "More than one row with the given identifier was found: " +
											        jdbcValuesSourceProcessingState.getExecutionContext()
													        .getEntityId() +
											        ", for class: " +
											        resultJavaType.getName()
							        );
						        }
						        add = false;
						        break;
					        }
				        }
					if ( add ) {
						results.add( row );
					}
					rowProcessingState.finishRowProcessing();
				}
			}
			else {
				while ( rowProcessingState.next() ) {
					results.add( rowReader.readRow( rowProcessingState, processingOptions ) );
					rowProcessingState.finishRowProcessing();
				}
			}
```

Reply:
done, Thanks @beikov for the review :+1: